gettsim/parameter_vizualisation_simon/bokeh_app/Scripts/wohngeld_interact.py

```diff
@@ -0,0 +1,169 @@
+# Import
+import numpy as np
+import pandas as pd
+from bokeh.layouts import column
+from bokeh.models import BasicTicker
+from bokeh.models import ColorBar
+from bokeh.models import ColumnDataSource
+from bokeh.models import LinearColorMapper
+from bokeh.models import NumeralTickFormatter
+from bokeh.models import Panel
+from bokeh.models import Slider
+from bokeh.palettes import Magma256
+from bokeh.plotting import figure
+from bokeh.transform import transform
+
+from gettsim import set_up_policy_environment
+from gettsim.transfers.wohngeld import wohngeld_basis
+
+
+def wohngeld_tab(plot_dict):
+    plot_dict = plot_dict["wohngeld"]
+
+    # Prepare wohngeld data ALTERNATIVE
+    def prepare_data(sel_year, hh_size):
```

Comment:
could your performance problems be related to the fact that you define functions within functions? And by calling this function, you overwrite the definition?

Reply:
I don't think my functions overwrite their definitions. Should I generally avoid defining functions inside functions? 

The performance issues are not new, they were there even before moving the code snippets into functions.

Take the kindergeld data preparation which is rather simple, the loop over all 45 years takes quite long. It is like this with any multi year data preparation. This might be a general issue with the approach of running the set_up_policy_environment to generate the data instead of having the data ready upon running the plots. What do you think?
![grafik](https://user-images.githubusercontent.com/74649653/103441034-4e9b7080-4c4b-11eb-86a9-ace171e0a41e.png)
`def prepare_data(start, end):
    """
    Data preparation for kindergeld parameters. Returns a dataframe.

    Parameters:
    start (Int): Defines the start of the simulated period
    end (Int):  Defines the end of the simulated period
    """

    # Calculate simulation period
    years = range(start, end)

    # Data preparation for Kindergeld params
    kindergeld_df = pd.DataFrame()

    for i in years:
        policy_params, policy_functions = set_up_policy_environment(i)
        kindergeld_df[i] = policy_params["kindergeld"]["kindergeld"].values()

    kindergeld_df = kindergeld_df.transpose()

    kindergeld_labels = ["First child", "Second child", "Third child", "Fourth child"]
    kindergeld_df.columns = kindergeld_labels

    return kindergeld_df`

Reply:
Without checking, I'd suspect that a nested `plotstyle()` is not accessible from another module. 

But if the slow performance is due to data processing, loading them once probably makes sense.